src/wallet/wallet.cpp

```diff
@@ -3010,14 +3024,18 @@ bool CWallet::CommitTransaction(CWalletTx& wtxNew, CReserveKey& reservekey, CCon
         // Track how many getdata requests our transaction gets
         mapRequestCount[wtxNew.GetHash()] = 0;
 
+        // Get the inserted-CWalletTx from mapWallet so that the
+        // fInMempool flag is cached properly
+        CWalletTx& wtx = mapWallet[wtxNew.GetHash()];
```

Comment:
Somewhat unrelated nit: AddToWallet's return value should be checked before using it here, but that was already the case.

Reply:
Indeed, though in such cases not much we can do - bdb got upset, we're probably just going to shutdown soon anyway.